stores/sql/main.go

```diff
@@ -1858,6 +1922,112 @@ func RemoveOfflineHosts(ctx context.Context, tx sql.Tx, minRecentFailures uint64
 	return res.RowsAffected()
 }
 
+// RenameDirectories renames all directories in the database with the given
+// prefix to the new prefix.
+func RenameDirectories(ctx context.Context, tx sql.Tx, bucket, prefixOld, prefixNew string) (int64, error) {
```

Comment:
Apart from the locking this PR looks pretty good now. If there was one thing I'd still nit about it's this function.

It feels like it could be reduced into 2 UPDATE queries where the first one updates all the parents for directories which would result in a conflict if renamed and the second one renames only the dirs for which it is safe to do so.

Achieving this would probably require selecting the dir name, removing the old prefix, adding the new prefix and then joining with the directories table to get the potential conflicts. So it might be a bit nasty. I'll leave it up to you whether you want to just try it out for fun to see if it cleans this up a bit. It definitely seems to have the potential to be faster than looping over each directory.

Reply:
Ok so I looked into this and wrote out the update statements that could make it work:

```sql
UPDATE directories AS child
JOIN directories AS d1 ON child.db_parent_id = d1.id
JOIN directories AS d2 ON d2.name = ? || SUBSTR(d1.name, ?))
SET child.db_parent_id = d2.id
WHERE d1.name LIKE ? AND d1.db_bucket_id = ?

UPDATE directories AS d1
LEFT JOIN directories AS d2 ON d2.name = ? || SUBSTR(d1.name, ?))
SET d1.name = ? || SUBSTR(d1.name, ?))
WHERE d1.name LIKE ?  AND d1.db_bucket_id = ? AND d2.id IS NULL
```

However, this would come at the cost of having to duplicate the code in `RenameDirectories` since the implementation would differ between SQLite and MySQL. For me that's not worth it, especially because the current version is easier to read and maintain. You mention it would be faster than looping over each directory but we only have to loop over a very limited subset of the directories, so performance shouldn't be an issue either. Agree or do you want to go for it?

Reply:
Benchmarks on my machine (Apple M1 Max):
```
old: BenchmarkRenameDirectories-10              21966             54057 ns/op           10418 B/op        251 allocs/op
new: BenchmarkRenameDirectories-10              10000            109526 ns/op           10007 B/op        241 allocs/op
```

I haven't committed the code yet because I have yet to clean up the statements:

```
// update db_parent_id
WITH cte AS (
	SELECT id, ? || SUBSTR(name, ?) as rename 
	FROM directories 
	WHERE name LIKE ?
)
UPDATE directories
SET db_parent_id = (
  SELECT d.id
  FROM directories d
  INNER JOIN cte ON d.name = cte.rename
  WHERE d.db_bucket_id = ?
  LIMIT 1
)
WHERE db_parent_id IN (SELECT id FROM cte) AND db_bucket_id = ?

// update name
WITH cte AS (
	SELECT id, ? || SUBSTR(name, ?) as rename 
	FROM directories 
	WHERE name LIKE ? AND name != ? AND db_bucket_id = ?
)
UPDATE directories SET name = (SELECT cte.rename FROM cte WHERE cte.id= directories.id  ) WHERE id IN (SELECT id FROM cte)
```

Turns out the new statements are slower... Now I'm a bit torn because I think the new statements I came up with are not optimal. I don't see how though, I gave it more than the good old college try. Regardless of the results I wouldn't go for duplicating the code and ending up with more complex queries. What do you think? Definitely don't want to give you the impression I'm throwing in the towel or something.

Reply:
I'd say it's definitely possible that the substring join is super slow compared to the iterative approach that always uses an index. Still great that we went the extra mile to confirm.